File: MergeSort.java
package ctci;
//https://www.hackerrank.com/challenges/ctci-merge-sort


import java.util.Scanner;

public class MergeSort {

	public static int mergeSort(int arr[]){
		return mergesort(arr,new int[arr.length],0,arr.length-1);
	}
	public static int mergesort(int[] arr, int temp[], int LeftStart, int RightEnd){
		int swaps = 0;
		if(LeftStart>=RightEnd){
			return 0;
		}
		int mid = (LeftStart + RightEnd) / 2;
		swaps = mergesort(arr, temp, LeftStart, mid);
		swaps += mergesort(arr, temp, mid+1, RightEnd);
		swaps+= merge(arr, temp, LeftStart, RightEnd);
		return swaps;
	}
	
	public static int merge(int[] arr, int[] temp, int leftStart, int rightEnd) {

		int swaps=0;
		int leftEnd = (leftStart + rightEnd) /2;
		int rightStart = leftEnd + 1;
		
		int size = rightEnd - leftStart + 1;
		int index = leftStart;
		int left = leftStart;
		int right = rightStart;
		while(left<=leftEnd && right<=rightEnd){
			if(arr[left]<= arr[right]){
				temp[index] = arr[left];
				left++;
			}
			else{
				temp[index] = arr[right];
				right++;
				swaps += (leftEnd + 1  - left);
			}
			index++;
		}
		
		System.arraycopy(arr, left, temp, index, leftEnd-left + 1);
		System.arraycopy(arr, right, temp, index, rightEnd - right + 1);
		System.arraycopy(temp, leftStart, arr, leftStart, size);
		return swaps;
	}
	public static void main(String[] args) {

		Scanner in = new Scanner(System.in);
		int t = in.nextInt();
		while(t-->0){
			int n = in.nextInt();
			int arr[] = new int[n];
			for (int i = 0; i < arr.length; i++) {
				arr[i] = in.nextInt();
			}
			System.out.println(mergeSort(arr));
			
		}

	}

}

File: mergeSort.py
#!/bin/python3

## https://www.hackerrank.com/challenges/ctci-merge-sort/problem

import sys
c=0

def countInversions(arr):
    # Complete this function
    #print(arr)
    global c
    c=0
    mergeSort(arr, 0, len(arr))
    return c
    
def mergeSort(arr, p, q):
    if(p<q):
        r = int(int(p+q)/2)
        if p!=r:
            mergeSort(arr, p, r)
        if r+1!=q:
            mergeSort(arr, r+1, q)
        #print (p,r,q)
        if r!=len(arr)-1 and arr[r]>arr[r+1]:
            merge(arr, p,r,q)

def merge(arr, p, r, q):
    s1=0
    s2=0
    global c
    arrl = arr[p:r+1]
    arrl.append(sys.maxsize)
    arrr = arr[r+1:q+1]
    arrr.append(sys.maxsize)
    arr2 = []
    ma = sys.maxsize
    while arrl[s1]!=ma and arrr[s2]!=ma:
        #print(s1,s2,arrl,arrr)
        if(arrl[s1] <= arrr[s2]):
            arr2.append(arrl[s1])
            s1+=1
        else:
            c+=len(arrl)-1-s1
            arr2.append(arrr[s2])
            s2+=1

    while s1 < len(arrl)-1:
        arr2.append(arrl[s1])
        s1+=1
    while s2 < len(arrr)-1:
        arr2.append(arrr[s2])
        s2+=1

    arr[p:q+1] = arr2
#    print(arr,c)
    return arr

if __name__ == "__main__":
    t = int(input().strip())
    for a0 in range(t):
        n = int(input().strip())
        arr = list(map(int, input().strip().split(' ')))
        result = countInversions(arr)
        print(result)
